src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/ReturningHandlerChain.cs

```diff
@@ -0,0 +1,164 @@
+        /// </param>
+        /// <param name="exceptionBehavior">
+        /// Behavior to observe when an exception occurs during execution of the chain. See
+        /// <see cref="ReturningHandlerChainExceptionBehavior"/> for explanation of how each option
+        /// affects the execution of the chain.
+        /// </param>
+        public ReturningHandlerChain(
+            ICollection<IReturningHandler<TParameters, TOutput>> handlers,
+            ReturningHandlerChainExceptionBehavior exceptionBehavior)
+        {
+            Debug.Assert(handlers is not null, "Collection of handlers must not be null.");
+
+            _exceptionBehavior = exceptionBehavior;
+            _handlerChain = handlers;
+        }
+
+        /// <summary>
+        /// Executes the chain of responsibility, while observing the desired exception behavior.
+        /// </summary>
+        /// <param name="parameters">Input to the chain of responsibility.</param>
+        /// <param name="isAsync">Whether async pathways should be used.</param>
+        /// <param name="ct">Cancellation token that will be passed to each handler.</param>
+        /// <returns>
+        /// If a handler could successfully handle the input, the result from that handler is
+        /// returned. If no handler could successfully handle the input, an exception is thrown.
+        /// </returns>
+        /// <exception cref="InvalidOperationException">
+        /// Thrown if an invalid exception behavior enum value is provided (which should never happen).
+        /// </exception>
+        public ValueTask<TOutput> Handle(TParameters parameters, bool isAsync, CancellationToken ct) =>
+            _exceptionBehavior switch
+            {
+                ReturningHandlerChainExceptionBehavior.Halt => HandleHalt(parameters, isAsync, ct),
+                ReturningHandlerChainExceptionBehavior.ThrowCollected => HandleThrowCollected(parameters, isAsync, ct),
+                ReturningHandlerChainExceptionBehavior.ThrowFirst => HandleThrowFirst(parameters, isAsync, ct),
+                ReturningHandlerChainExceptionBehavior.ThrowLast => HandleThrowLast(parameters, isAsync, ct),
+                _ => throw new InvalidOperationException(),
+            };
+
+        private async ValueTask<TOutput> HandleHalt(TParameters parameters, bool isAsync, CancellationToken ct)
+        {
+            foreach (IReturningHandler<TParameters, TOutput> handler in _handlerChain)
+            {
+                // If handle throws an exception, we let it bubble up higher.
+                TOutput result = await handler.Handle(parameters, isAsync, ct);
+                if (result is not null)
+                {
+                    return result;
+                }
+            }
+
+            // If we make it to here, none of the handlers successfully handled the message. Throw
+            // that no handlers found.
+            throw new NoSuitableHandlerFoundException();
```

Comment:
Note that this should not lead to public exception, should be caught and handled where expected.

Reply:
Yes, this is something that will likely be addressed in a later PR.